Sources/SwiftProtobuf/JSONScanner.swift

```diff
@@ -706,6 +398,311 @@ internal struct JSONScanner {
+          // Slow path: JSON allows floating-point notation for integers
+          index = start
+          if let d = try parseBareDouble(source: source,
+                                         index: &index,
+                                         end: end) {
+            if let u = UInt64(exactly: d) {
+              return u
+            }
+          }
+          throw JSONDecodingError.malformedNumber
+        case asciiBackslash:
+          return nil
+        default:
+          return n
+        }
+      }
+      return n
+    case asciiBackslash:
+      return nil
+    default:
+      throw JSONDecodingError.malformedNumber
+    }
+  }
+
+  // Parse the leading Int64 from the provided utf8.
+  //
+  // This uses parseBareUInt() to do the heavy lifting;
+  // we just check for a leading minus and negate the result
+  // as necessary.
+  //
+  // As with parseBareUInt(), if it encounters a "\" backslash
+  // character, it returns a nil.  This is used by callers that are
+  // parsing quoted numbers.  See nextSInt() and nextUInt() below.
+
+  private func parseBareSInt(
+    source: UnsafeBufferPointer<UInt8>,
+    index: inout UnsafeBufferPointer<UInt8>.Index,
+    end: UnsafeBufferPointer<UInt8>.Index
+  ) throws -> Int64? {
+    if index == end {
+      throw JSONDecodingError.truncated
+    }
+    let c = source[index]
+    if c == asciiMinus { // -
+      source.formIndex(after: &index)
+      // character after '-' must be digit
+      let digit = source[index]
+      if digit < asciiZero || digit > asciiNine {
+        throw JSONDecodingError.malformedNumber
+      }
+      if let n = try parseBareUInt(source: source, index: &index, end: end) {
+        let limit: UInt64 = 0x8000000000000000 // -Int64.min
+        if n >= limit {
+          if n > limit {
+            // Too large negative number
+            throw JSONDecodingError.numberRange
+          } else {
+            return Int64.min // Special case for Int64.min
+          }
+        }
+        return -Int64(bitPattern: n)
+      } else {
+        return nil
+      }
+    } else if let n = try parseBareUInt(source: source, index: &index, end: end) {
+      if n > UInt64(bitPattern: Int64.max) {
+        throw JSONDecodingError.numberRange
+      }
+      return Int64(bitPattern: n)
+    } else {
+      return nil
+    }
+  }
+
+  // Identify a floating-point token in the upcoming UTF8 bytes.
+  //
+  // This implements the full grammar defined by the JSON RFC 7159.
+  // Note that Swift's string-to-number conversions are much more
+  // lenient, so this is necessary if we want to accurately reject
+  // malformed JSON numbers.
+  //
+  // This is used by nextDouble() and nextFloat() to parse double and
+  // floating-point values, including values that happen to be in quotes.
+  // It's also used by the slow path in parseBareSInt() and parseBareUInt()
+  // above to handle integer values that are written in float-point notation.
```

Comment:
Also calling out why backslash returns nil here would help, when first reading the code, it seemed like these should be errors, not return nils.